### ISSUE TITLE
native histograms: Implement the < and > operators to "trim" a histogram.

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#14651]
### Proposal

This feature was now requested twice independently in different contexts, so I thought we should have an issue for it.

The short story is to implement the `>` and `<` operators between a native histogram and either a scalar or a vector with floats. The operators will return a histogram with observations greater or less than the float value removed. (This will be an estimate in most cases and not precise.)

The more recent feature discussions happened [on the dev mailing list](https://groups.google.com/g/prometheus-developers/c/Zmk8ezTot4o). 

Quoting from there:

[The idea is] to "simply" implement the
`>` and `<` operator between native histograms and scalars/floats. It
would return a new histogram with all the observations below or above
the threshold given by the scalar removed. (This will be an estimate
in most cases, but given the generally high resolution of native
histograms, that's OK.)

In that way, you can already "trim" a histogram at a given
threshold. This will return a histogram containing only requests that
lasted longer than 100ms:

    request_duration_seconds > 0.1

You can then combine this with other PromQL expressions to implement
trimming at percentages. The following will exclude the 25% shortest
requests:

    request_duration_seconds > histogram_quantile(0.25, request_duration_seconds)

From there, you can use all the other tools to do something with the
returned histogram, e.g. calculating a mean or a median or whatever
you want.

**Comment by juliusv** [comment]
Makes sense. It's interesting to note that so far, comparison operators have applied to sample values, which in classic histograms would have been the bucket counts and not the bucket bounds (which were part of the time series identity). So if one stuck closer to the existing behavior of comparison operators, `rate(request_duration_seconds[5m]) > 1` would mean, "Give me a derived histogram with only the buckets that have a rate of more than 1 per second". I just wanted to mention this so that it's at least considered. I can imagine some use cases for either case, but I'm not sure which one would be more relevant. Just that comparing to the bucket count vs. bounds would be more in line with the existing behavior of comparison operators.

**Comment by krajorama** [comment]
I think using `<` or `>` this way would be very confusing as the existing compare operators mean: do not give me this sample if the compare is false.

So `rate(request_duration_seconds[5m]) > 1` is easily understood as : only give me results here if all observations are higher than 1 unit.

I'd rather we had a different operator or a function.
So for example:
`ltrim(1, rate(request_duration_seconds[5m]))`  meaning lower (left) trim, and conversly `utrim` for upper trim and you could have `trim` to be able to do both.

Or maybe some operator `<<` or `<-` ?

**Comment by beorn7** [comment]
As a function, we could also have `histogram_trim(lower scalar, upper scalar, v instant-vector)`, mirroring the same signature as `histogram_fraction`. My worry with the function approach is that one of the prime use cases of this is to trim a histogram by quantile values, which you would calculate with `histogram_quantile`, and then you get a vector as output and really want vector matching, which (IIRC) you only get with operators but never with functions in PromQL.

My current believe is that this means we really want an operator for the trimming rather than a function, so we should discuss the operator syntax, but if anyone has more arguments in favor of a function, please let us know.

**Comment by beorn7** [comment]
To summarize the concerns about the operator syntax (see above, but also from personal communication):
1. `>` and `<` in the conventional setting are _filtering_, i.e. they either return the LHS unchanged or they do not return anything. The operation proposed here would _mutate_ the histogram, which is a very different thing (although one might argue we are still _filtering_ observations out of the original histogram and we return a filtered histogram…). A user might expect a more direct filtering behavior, e.g. `rate(request_duration_seconds[5m]) > 0.1` could be read as "return the histogram unchanged if it contains only observations greater than 0.1".
2. The meaning of `>` and `<` for classic histograms is very different: `rate(request_duration_seconds_buckets[5m]) > 0.1`  returns only buckets with a population greater than 0.1, while `rate(request_duration_seconds[5m]) > 0.1` (with the proposal here) would return a histogram with only observations of greater 0.1.
3. The confusion gets even greater if we have a mixed vector of float samples and histogram samples, because then the different behavior would mix up in the same expression.

I think these are valid concerns, and finding a better symbol for the operator would be great. But what symbol could we use? `<<`/`>>` is used for bitshift operations (or some kind of pipelining in C++), `<-`/`->` is used for channels in Go and for pointers in C. My concern is that they might just cause even more confusion.

Playing devil's advocate, here an attempt to defuse the above concerns (same item numbering):
1. The main argument here is already hinted above: We are still filtering, just in a different way. Given the very different nature of histogram samples vs. float samples, the differences are maybe not too surprising. The remaining question is if the suggested understanding of "return the histogram unchanged if it contains only observations greater than one" is really something that users will intuitively come up with. If this feature were useful, it would also be a strong argument that users will assume the operator works that way. But is it really useful? In general, I think operators acting on histograms is inherently complex, and users won't necessarily have an intuition at all, implying that it's fine to let them learn from the documentation.
2. How the bucket series are handled is mostly emergent from the way PromQL works in general. It's not generally a good role model, especially if the original query doesn't make a lot of sense. (I deliberately picked 0.1 as the threshold above because that let's the query appear even more nonsensical. ;)
3. While it _is_ really confusing if you actually have a mixed-type vector, it is in practice extremely rare to encounter mixed-type vectors. It will probably only happen during some metrics renaming transition, so the relevance for our operator pick is limited.

In summary, I would be happy with a different operator symbol, as long as it isn't even more confusing than the confusion caused by `<`/`>` as discussed above.

**Comment by krajorama** [comment]
Some ideas:

there are already operator words : `and` `or` `unless`. What would be a good word for this?

None of these look really appealing , having a short operator looks more natural, readable, even if it's two characters.
`0.1 trim rate(request_duration_seconds[5m]) trim 0.9` 
`0.1 histogram_trim  rate(request_duration_seconds[5m]) histogram_trim 0.9` 
`0.1 bound rate(request_duration_seconds[5m])  bound 0.9`

I don't have a better idea currently than `</` . Would solve my SLO use case:
`histogram_count(sum(rate(request_duration_seconds[5m] </ 2.5)))`
or indeed
`sum(histogram_count(rate(request_duration_seconds[5m] </ 2.5)))`

**Comment by beorn7** [comment]
Yes, a word as an operator is also an option. To me, they kind of look weird (not just in this case). `and` and `or` are kind of "well known" as operators from other languages. `unless` is at least a conjunction in English. `atan2` is definitely in weird-land as it looks like it should be a function (and we made it a operator so we could do label matching).

The perfect fitting word for this, if we can find one, would be a nice solution.

**Comment by krajorama** [comment]
To add more motivation to do this: in SLO scenario a typical calculation is to find successful requests that took less than a time limit (let's say 5 seconds). Which now you write as:

histogram_fraction(0, 5, sum(rate(request_duration_seconds{state=success}[5m])))
*
histogram_count(sum(rate(request_duration_seconds{state=success}[5m])))

This is a bit ugly, which is one thing, but also if the series of request_duration are all idle (not increasing counters), the `rate` of a constant histogram will be 0. Which means the `sum` will be 0 as well. However `histogram_fraction` is **NaN** by [definition](https://github.com/prometheus/prometheus/blob/c586c15ae6a0fdb4d02abb9b850bcfedb44644e9/promql/promqltest/testdata/native_histograms.test#L812) on 0 histograms.

This makes sense mathematically, however doesn't answer the question of how many requests were successful and took less than 5 seconds. Which I would expect to be 0.

Note that is I just ask for the successful requests:
```
histogram_count(sum(rate(request_duration_seconds{state=success}[5m])))
```
then this does give me the correct 0 result even if the sum of rates was 0.

Conversely if we had for example the syntax 
```
histogram_count(sum(rate(request_duration_seconds[5m] </ 2.5)))
```
this would give the expected 0 count as it does not involve fractions/quantiles, thus no divisions.

**Comment by beorn7** [comment]
This looks good in general, we "only" need to make a call how to name the operator.

By now, I'm pretty sure we shouldn't use the `<` and `>` directly, for all the arguments above, but also I think we don't really need the different directions. I think we need only one operator, and the order of operands will tell if it is a "left trim" or a "right trim". Then we need one fewer operator, and I think the flow of reading makes most sense now, see examples later.

We still have to decide if we want a "word" operator (current proposal: `trim`) or a "symbolic" operator (current proposal `</`). With no other proposals so far, I'll just write down some example queries and then we can have a straw poll via emoji reactions…

The examples assume there is already a recording rule like 
```YAML
record: request_duration_seconds:rate5m
expr: sum(rate(request_duration_seconds[5m]
```
```

**Comment by beorn7** [comment]
# Symbolic operator

Histogram with only observations less or equal 2.5:
```PromQL
request_duration_seconds:rate5m </ 2.5
```

Histogram that excludes the bottom 10% and top 10% of observations:
```PromQL
histogram_quantile(0.1, request_duration_seconds:rate5m) </ request_duration_seconds:rate5m </ histogram_quantile(0.9, request_duration_seconds:rate5m)
```

The same, but on multiple lines:
```PromQL
  histogram_quantile(0.1, request_duration_seconds:rate5m) 
</ 
  request_duration_seconds:rate5m 
</ 
  histogram_quantile(0.9, request_duration_seconds:rate5m)
```

If you like or dislike this syntax, use corresponding emoji reactions on this comment.

**Comment by beorn7** [comment]
# Word operator

Histogram with only observations less or equal 2.5:
```PromQL
request_duration_seconds:rate5m trim 2.5
```

Histogram that excludes the bottom 10% and top 10% of observations:
```PromQL
histogram_quantile(0.1, request_duration_seconds:rate5m) trim request_duration_seconds:rate5m trim histogram_quantile(0.9, request_duration_seconds:rate5m)
```

The same, but on multiple lines:
```PromQL
  histogram_quantile(0.1, request_duration_seconds:rate5m) 
trim 
  request_duration_seconds:rate5m 
trim 
  histogram_quantile(0.9, request_duration_seconds:rate5m)
```

If you like or dislike this syntax, use corresponding emoji reactions on this comment.

**Comment by krajorama** [comment]
It's just hard to imagine anything as intuitive as `</` .
But I was wondering about the effect of typos:
- extra space in the middle: `< /` which would fail parsing I would hope.
- only writing `<` which could mean filtering, a totally valid use case.
- only writing `/`  which would mean division, a totally valid use case.

Still I think `</` is just so much more readable that I'll still vote that.

**Comment by zenador** [comment]
I find word operators more readable and easier to search for more info to understand what it does, but that's just personal preference. If it's not a simple arithmetic or comparison operator that are well-known and intuitive, people coming across it for the first time would be confused and would need to look it up somehow, so words would be easier.

**Comment by ellisda** [comment]
The `rate5m` examples read well with the symbolic. I'm pretty naive here, but I was surprised / confused by where it was in the raw form example:
* `histogram_count(sum(rate(request_duration_seconds[5m] </ 2.5)))`

Something about bringing the operator inside the `[5m]` range feels very odd. I guess if you can use the operator on the result of the `:rate5m` rule, you'd also be able to use in the same place in the raw expression: `histogram_count(0.1 </ sum by(...) (rate(request_duration_seconds[5m])) </ 0.9)` right? Would it work both inside AND outside the instant vector?

**Comment by juliusv** [comment]
I don't know yet which syntax I prefer, but I think I'd want to have two operators. When it's a word like `trim`, I'll have trouble remembering which way it trims.

But even in this example with `</`:

```
  histogram_quantile(0.1, request_duration_seconds:rate5m) 
</ 
  request_duration_seconds:rate5m 
</ 
  histogram_quantile(0.9, request_duration_seconds:rate5m)
```

This would contradict the existing pattern where binary operators always return the sample value from the left (first) operand, not the right (second) one. The `or` operator is the only exception to that, but that's only because that is basically its entire purpose.

So I think having two operators is still a better way to go. Then it'd be:

```
  request_duration_seconds:rate5m
>/ 
  histogram_quantile(0.1, request_duration_seconds:rate5m) 
</ 
  histogram_quantile(0.9, request_duration_seconds:rate5m)
```

Sticking to this is not only more consistent with existing binary operators, but it also allows you to put the most important thing (the histogram itself) first, where you can more easily see and read it.

**Comment by juliusv** [comment]
> The `or` operator is the only exception to that, but that's only because that is basically its entire purpose.

Just to add and clarify: I was talking about vector-to-vector binops here. You can write a scalar-to-vector `1 > foo` expression (although super uncommon), and the sample value will come from the right. That shouldn't be the case for two vectors though.

**Comment by beorn7** [comment]
Thanks, @juliusv . That's a good point. I think given the different nature of the new operator, we _could_ introduce this different kind of behavior. It would be well defined, but it would actually need quite a lot of words to define it well. (I was already worried about how the documentation would read for something that ultimately is quite easy to use).

So yeah, currently I tend toward implementing both `>/` and `</` (although I liked to implement only one of them previously), and strictly require the histogram to be first and the float (from vector or scalar) to be second.

To do the same with a "word" operator, we needed a longer word, something like `trimleft` and `trimright` or `trimlower` and `trimupper`. I guess that changes the balance even more towards a "symbol" operator.

**Comment by juliusv** [comment]
> and strictly require the histogram to be first and the float (from vector or scalar) to be second.

Since we currently allow `42 > my_vector`, it would seem odd if there are other comparison/filtering ops which suddenly don't allow it. Although I don't think it would affect many real-world uses.

**Comment by beorn7** [comment]
One argument is that it isn't comparing and it isn't filtering. It is removing observations from a histogram, returning a mutated histogram. That's kind-of the argument that @krajorama brought up: We shouldn't use `<` and `>` for something that works so differently.

**Comment by krajorama** [comment]
Agree with @juliusv after reading  [arithmetic binary operators](https://prometheus.io/docs/prometheus/latest/querying/operators/#arithmetic-binary-operators). Also it doesn't affect my use case of just trimming for less than 2s (const scalar) buckets.

The one thing that could wrong is typos (people writing < or / only) and possibly some slight confusion whether the operator is in arithmetic binary or comparison binary category , but I guess we'll just document it.

**Comment by beorn7** [comment]
Related to this (but triggered by an unrelated issue), @krajorama, @zenador, and myself discussed @krajorama's earlier idea about giving the `<`, `>`, `<=`, and `>=` operators a useful behavior if applied to a native histogram on the one side and a float on the other side (both for floats in vector elements or as scalars). They would work as filters as usual (or returning 0 or a with the `bool` modifier), but the condition is referring to the observations in the histogram.

For example `rpc_duration_seconds < 0.1` (or equivalently `0.1 > rpc_duration_seconds` returns nothing for histogram elements in `rpc_duration_seconds`  that have populated buckets for observations ≥ 0.1, or in other words where all populated buckets only count observations < 0.1. With `<=` / `>=`, the inclusion rules are adjusted correspondingly (but it only matters if the threshold coincides with a bucket boundary). Also, `rpc_duration_seconds > 0.1` works correspondingly in inverted fashion. With floats from a vector, the order matters to determine if the float or the histogram is returned if the condition is true.

Note that there are subtle differences in semantics. `float_metric <= 5` will return the inverted set of elements compared to `float_metric > 5` (with the notable exception of `NaN` elements, which will not be returned by either expression). However `histogram_metric <= 5` and `histogram_metric > 5` will only work in the same way if there is a bucket with an upper inclusive boundary of 5. Histograms for which 5 is in the middle of a populated bucket won't be returned by either expression.

What is a bit problematic is the (lack of) consistency with the proposed `</` and `>/` operators (which is also the reason why we thought we should discuss this proposal as part of this issue first before elevating it to its own proposal). If I do `histogram_metric </ 0.1`, I expect to get a histogram where all observations greater than 0.1 have been removed. However, `(histogram_metric </ 0.1) < 0.1` will return nothing in the (likely) case that 0.1 is in the middle of a populated bucket. The trimming will reduce the count in that bucket via interpolation, but the bucket will still be populated, so the comparison operator has to assume there are still (possibly) observations ≥ 0.1 in the histogram. Things will only work out as (mathematically) expected if there is a bucket boundary precisely at 0.1 (or if the bucket accepting 0.1 happens to be unpopulated in the original histogram).

**Comment by beorn7** [comment]
Hello from the bug scrub!

This is being worked on.

**Comment by manmathbh** [comment]
Hi!  I'd like to work on implementing the `<` and `>` operators for native histogram trimming.

I've been contributing to Prometheus (see #17882 - YAML multi-document warning) and would love to tackle this feature.

My plan:
1. Study the existing native histogram operators in PromQL
2. Implement the trimming logic for `<` and `>` operators  
3. Add comprehensive tests
4. Update documentation

Would this be okay? Any guidance on where to start would be appreciated! 

@beorn7

**Comment by juliusv** [comment]
@manmathbh FYI there's an open but stale PR for this open at https://github.com/prometheus/prometheus/pull/16330. As the last recent comment on it, @linasm offered to drive it past the finish line: https://github.com/prometheus/prometheus/pull/16330#issuecomment-3753348097

**Comment by linasm** [comment]
> [@manmathbh](https://github.com/manmathbh) FYI there's an open but stale PR for this open at [#16330](https://github.com/prometheus/prometheus/pull/16330). As the last recent comment on it, [@linasm](https://github.com/linasm) offered to drive it past the finish line: [#16330 (comment)](https://github.com/prometheus/prometheus/pull/16330#issuecomment-3753348097)

Yeah, I'm already working on it.

**Comment by manmathbh** [comment]
> [@manmathbh](https://github.com/manmathbh) FYI there's an open but stale PR for this open at [#16330](https://github.com/prometheus/prometheus/pull/16330). As the last recent comment on it, [@linasm](https://github.com/linasm) offered to drive it past the finish line: [#16330 (comment)](https://github.com/prometheus/prometheus/pull/16330#issuecomment-3753348097)

Got it. Thanks @juliusv and @linasm!

**Comment by krajorama** [comment]
From discussions in the https://github.com/prometheus/prometheus/pull/17904 PR : 

* The spec should say that `histogram </ x` remove observations above `x` and keeps those below. Text above to be udpated.
* The spec should say that the `histogram_sum` value after the trim operator is an estimate that can be widely inaccurate.

Example for the `hisotgram_sum` being off:

Imagine a wide bucket (e.g. custom bucket or low resolution exponential histogram)  where the observations happened at near the edge of the bucket. We're assuming even linear/exponential distribution of observations in the bucket when we calculate the cutoff point inside the bucket and calculate the change in the `histogram_sum` based on that. But reality can be very different.
